pkgsStatic: make pkgsMusl a no-op

r-burns opened the pull request:
When pkgsStatic is already musl, we can let pkgsMusl simply be a no-op.

As the interaction of pkgsStatic and pkgsMusl machinery is currently broken
(see discussion in https://github.com/NixOS/nixpkgs/issues/114510), this fixes pkgsStatic.pkgsMusl.stdenv.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

r-burns committed:
pkgsStatic: make pkgsMusl a no-op

When pkgsStatic is already musl, we can let pkgsMusl simply be a no-op.
As the interaction of pkgsStatic and pkgsMusl machinery is currently
broken, this fixes pkgsStatic.pkgsMusl.stdenv.

r-burns commented:
Still relevant. Previous command gives same result on current master.

rnhmjoj commented:
The fact `pkgsStatic` is using musl is an accident: glibc static build was broken at the time and musl was used instead. Ideally `pkgsStatic should differ from `pkgs` in just having `-static` kind of flags.

Nowadays `glibc.static` is working and I was actually thinking about changing `pkgsStatic` back to using it.

r-burns commented:
Well, this whole contraption is conditional on isMusl anyway, so if we stop setting useMusl for glibc linux, it should gracefully use static glibc instead.

superherointj commented:
@rnhmjoj Do you still plan on moving static back to glibc? (I guess would be nice if you did it. It is quite odd the way it is now. It took me a while to understand why static was using musl. / I was out of context.)

rnhmjoj commented:
That's the idea, but I don't plan on working on this soon.

yu-re-ka commented:
I don't think `pkgsMusl.pkgsStatic` should be a no-op compared to `pkgsStatic`. It changes the build platform from glibc to musl as well, so makes the difference if you are cross-compiling from glibc to musl-static, or from musl to musl-static.

r-burns commented:
Was looking through my old open PRs and discovered that this is *still* (!) an issue.

On master:

```
nix eval -f . pkgsStatic.pkgsMusl.stdenv.hostPlatform.isStatic                                                                                                                             
false
```

Rebased to fix conflicts, and simplified the patch a bit. There are no longer any assumptions about pkgsStatic using musl - the no-op overlay for pkgsMusl is only applied when `super'.stdenv.hostPlatform` is already Musl.

Would appreciate a review to fix this long-standing issue.

ConnorBaker reviewed (approved):
Seems entirely reasonable, I had to do something similar for the CUDA package sets (use the already-constructed package set rather than creating a new one to avoid clobbering/blowing away changes).
